### ISSUE TITLE
legacy syntax to disable with Oil

### ISSUE THREAD

**Issue by andychu** [oilshell/oil#678]
- `[[ ]]` -- replace by limited uses of `test`, and `globmatch()` function for `==`, etc.
- `(( ))`  and all shell arithmetic
- most of `${}`
- backticks -- or are we going to use this for something else?
- assignment builtins?   
  - except for `export`, I suppose
- aliases already mentioned in #677

**Comment by andychu** [comment]
Related to #1016

**Comment by andychu** [comment]
We also should disable subshells with `( )` syntax.

    $ bin/oil -c '( echo hi )'
    echo hi

I was unsure whether to make this part of `parse_paren`, because I wanted to break as little as possible in `oil:basic`.

So I suppose we could have `parse_subshell` which is turned off?    The preference is to use `forkwait` instead

**Comment by bar-g** [comment]
I find modes kind of unclear, or not self-explaining. How about...

oil:supplement -- oil:precedence -- oil:only

oil:allow -- oil:full -- oil:only?

/bin/oil using oil:only?

**Comment by andychu** [comment]
I kind of like the name `oil:only` ... Though right now I think it's good to keep it to 3 groups: `oil:basic`, `strict:all`, and `oil:all`, corresponding to the four use cases.

It's not clear that we want anything even more strict than `oil:all` -- what would be in `oil:only` and not `oil:all`?  

`oil:all` should disable everything we want disabled, I think.

Though it also relates to the versioning  #973   although I don't think we need that yet either

Most likely we will allow some stuff simply because it takes a lot of code to disallow it ... like all the stuff within `${}`

**Comment by andychu** [comment]
I think a better option is to have a separate lint tool that disallows even more at parse time.  So it's not even a shell options.

That would be related to #936  -- e.g. Python lint can warn you about undefined references

**Comment by bar-g** [comment]
Yep, if the implementation not close to, or it will never have all shell stuff disabled then it doesn't make much sense to rename `oil:all` to `oil:only`.

I would still see an improvement in renaming `oil:basic` to  `oil:allow`, though. This would actually make it obvious that it is not a breaking change. I can be deduced that shell compatible syntax is still valid, because oil is just "allowed" in addition. (I also first thought `oil:basic` could only be enabled as a next step after `strict:all`, or would include it, whereas `oil:allow` literally says it's just an orthogonal allowance, correct?)

And contrasting `oil:allow` it would be possible to convey a bit more that `oil:all` disables/break some parts of compatible shell by renaming it to `oil:full` (expands to its fullest extent).

And then mentioning the available set of options to the user as "`oil:allow`, `strict:all` and `oil:full`" (in that order) may even better (and correctly) convey that the last includes all stricts.

May I add this to #1127 ?

**Comment by andychu** [comment]
Hm I can see some argument for `oil:allow`, but it's not obviously better ... I guess I'd like to get the opinion of others

I'm wary of changing it now because people are using it.  It would have to be a big  improvement to be worth the breakage for a cosmetic thing

Though another option  is to name it after what you want to do:


```
shopt --set oil:improve  # improve existing script with "strict"
shopt --set oil:upgrade  # upgrade an existing script
shopt --set oil:new  # writing a new script
```

But yeah probably not worth the breakage ...

**Comment by andychu** [comment]
This issue is also related to #468  -- OSH / Oil subsets

We might want to close it in favor of that ... I'm basically just disabling legacy syntax "on demand", e.g. when I noticed the problem with `((` which is confusing

But if there's no problem, I'm not inclined to disable it IN THE INTERPRETER even if the syntax is ugly/inconsistent.  I think that is better left for tools

**Comment by bar-g** [comment]
Yes, good point to rather define optionally restricting subsets in parser options. And agreed not worth a breakage when it doesn't truly help new users.

Concerning oil:improve, oil:upgrade and oil:new. it might easily be misunderstood to improve or upgrade the oil language. That could be mitigated by partly using osh there, which could also make it more obvious that the last oil option includes those of osh:

`osh:upgraded`, `osh:strict` and `oil:full`

But, shorter and keeping more of their particular language/interpreter option group focus (this and an aw:full pun kept inking me about oil:full) could be this:

`oil:allow`, `osh:strict`, `oil:all`
Maybe that could help new users by just allowing the first two alternatives and using these in the docs, without breaking the old oil:basic and strict:all, yet.

I think the second triple (even in another order) could have made it pretty much clear to me from the beginning. 
If you agree I could add that to https://github.com/oilshell/oil/issues/1127 to gather feedback there.

**Comment by bar-g** [comment]
oh, the telling notions may be...

"this script is written with `oil:allow` and `osh:strict`"

Instead of "with `strict:all` and in `oil:basic`" (immediate question: what are these exactly?)

**Comment by bar-g** [comment]
Perspectively, may `oil:all` coming tomean `oil:strict`?

**Comment by andychu** [comment]
Hmmm ... I actually like `osh:strict`.  Because it contains the difference between OSH and Oil.  So it could be

`osh:strict`, `oil:basic`, `oil:all`

Then you sort of know what language you're writing in.  (Although note that `var` `const` `proc` are ALWAYS allowed regardless of option!)

(Not so keep on `oil:allow` since allow is a verb, not a noun.  Also what it does is **CHANGE** the meaning of `()` and `{}` most significantly.)

Hm yeah I'm wary of breaking things, but `osh:strict` is possibly more descriptive and consistent

**Comment by bar-g** [comment]
Uh, I'm surely not perfectly aware of all the changes.

I had thought of the expressions and blocks that get "allowed", for example, in `if (...) {...}`, and apparently have had wrongly assumed that explicit sub-shells could continue to work in other contexts. Same with `{...} >/dev/stderr`.

The working `var` `const` `proc` in osh could be attributed again to consistently prefer linter in the direction of restrictions again. (Allow as much incremental code refactoring as possible.) Wouldn't see that too much contradicting with `oil:allow`, but maybe we just can still find something that explains it better, here.

`osh+oil`, `osh:oil`, `oil:in-osh`, `osh:just-very-well-oiled` :-)  ...???

**Comment by andychu** [comment]
No actually subshells  `( )` and `{ }` continue to work, even in `oil:all`, regardless of options.

However they are not idiomatic any more.  New idioms are:

```
forkwait {
  echo one
  echo two
}

fopen > out.txt {
  echo one
  echo two
}  # I JUST added this, it's not even out yet
```

In a theoretical world we could figure out exactly what Oil is and disallow everything else.  In practice it's a compromise, and some people arrive to the party 10 years later than others.

Python and Ruby both have a lot of "old style stuff" that people don't use anymore.  Python 3 dropped a few things after ~18 years or so.


-----

I like the `osh:strict` suggestion, we can discuss on #1137.  I think it's doable without too much disruption

**Comment by bar-g** [comment]
Ok, now that I thought of the simple word eval I think *that* change could warrant prefering `oil:basic` over `oil:allowed`. The {} and () would perfectly weigh as "allowed" syntax IMHO.

Where is that party you get all the old stuff from? Thought, you were already past wanting to hard-disable everything but oil syntax above. :-)
